ADAL/src/workplacejoin/mac/ADWorkPlaceJoinUtil.m

```diff
@@ -22,14 +22,292 @@
+{
+    OSStatus status = noErr;
+    ADAuthenticationError* adError = nil;
+    NSData* issuer = nil;
+    NSMutableDictionary *identityQuery = nil;
+    CFDictionaryRef result = NULL;
+    
+    *identity = nil;
+    *clientCertificate = nil;
+    if (error)
+        *error = nil;
+
+    *clientCertificate = [self copyWPJCertificateRef:correlationId error:&adError];
+    
+    if (adError)
+    {
+        if (error)
+            *error = adError;
+        
+        AD_LOG_ERROR_F(@"Failed to retrieve WPJ client certificate from keychain - ", adError.code, nil, @"%@ correlation id.", correlationId);
+        goto _error;
+    }
+    
+    // If there's no certificate in the keychain, adError won't be set since this isn't an error condition.
+    if (!*clientCertificate)
+        return;
+    
+    // In OS X the shared access group cannot be set, so the search needs to be more
+    // specific. The code below searches the identity by passing the WPJ cert as reference.
+    identityQuery = [[NSMutableDictionary alloc] init];
+    SAFE_ARC_AUTORELEASE(identityQuery);
+    
+    [identityQuery setObject:(__bridge id)kSecClassIdentity forKey:(__bridge id)kSecClass];
+    [identityQuery setObject:(__bridge id)kCFBooleanTrue forKey:(__bridge id<NSCopying>)(kSecReturnRef)];
+    [identityQuery setObject:(__bridge id)kCFBooleanTrue forKey:(__bridge id<NSCopying>)(kSecReturnAttributes)];
+    [identityQuery setObject:(__bridge id)kSecAttrKeyClassPrivate forKey:(__bridge id)kSecAttrKeyClass];
+    [identityQuery setObject:(__bridge id)*clientCertificate forKey:(__bridge id)kSecValueRef];
+    
+    status = SecItemCopyMatching((__bridge CFDictionaryRef)identityQuery, (CFTypeRef*)&result);
+    CHECK_KEYCHAIN_STATUS(@"Failed to retrieve WPJ identity from keychain.");
+    
+    issuer = [(__bridge NSDictionary*)result objectForKey:(__bridge id)kSecAttrIssuer];
+    if (issuer)
+    {
+        *clientCertificateIssuer = [[NSString alloc] initWithData:issuer encoding:NSISOLatin1StringEncoding];
+        SAFE_ARC_AUTORELEASE(*clientCertificateIssuer);
+    }
+    
+    *identity = (__bridge SecIdentityRef)([(__bridge NSDictionary*)result objectForKey:(__bridge id)kSecValueRef]);
+    if (*identity)
+        CFRetain(*identity);
+    
+    CFRelease(result);
+    
+    return;
+    
+_error:
+    
+    if (*identity)
+        CFRelease(*identity);
+    *identity = nil;
+    
+    if (*clientCertificate)
+        CFRelease(*clientCertificate);
+    *clientCertificate = nil;
+    
+    *clientCertificateIssuer = nil;
+}
+
+
++ (SecCertificateRef) copyWPJCertificateRef:(NSUUID*)correlationId
+                                      error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    OSStatus status= noErr;
+    SecCertificateRef certRef = NULL;
+    NSData *issuerTag = [self wpjCertIssuerTag];
+    
+    NSMutableDictionary *queryCert = [[NSMutableDictionary alloc] init];
```

Comment:
We cut releases from dev on a regular basis. Because of that anything functionality added to dev has to be in a reasonably shippable state, capable of being fully tested and with minimal expectation of future breaking changes, or hidden in a way that it does not interfere with existing consumers, and aligned with the release schedule. What you're saying here is that there are future breaking changes expected to happen that will create incompatibilities in the future. To me that means this change should not be in 'dev' until we settle on those expected changes and aligns correctly with the release schedule.